#568 event auditor

joncppl opened the pull request:
Notes:

-used a bit of component state in EventAuditor/index.js, mostly to keep the count of is_submission events correct without fetching the list again / writing a custom reducer.

joncppl committed:
implement the ui for event auditor

joncppl committed:
improve display of event in audit page

-include event image
-improve readability of event fields

joncppl committed:
add functionality to event auditor

joncppl committed:
formatting

joncppl committed:
add pagination to event auditor

joncppl committed:
automatically change pages when one is cleared

joncppl committed:
reset event card state on new props

joncppl committed:
Merge branch 'Events' into #568-event-auditor

joncppl committed:
improve style of datetimeinput

psiemens commented on dispatch/apps/api/views.py:
```diff
@@ -403,6 +404,10 @@ def get_queryset(self):
             )
+        if pending:
+            queryset = queryset.filter(
+                Q(is_submission__exact=1)
```
You don't need to use `Q` here since there's only one condition. Also, the default queryset should exclude submitted events:

```py
if pending:
    queryset = queryset.filter(is_submission=True)
else:
    queryset = queryset.filter(is_submission=False)
```

Can you add a test case for this as well?

joncppl replied:
looks like axel has a test in Event-forms-etc. 
When that's merged I'll pull it into this branch.

Axel-Jacobsen replied:
I completed this in my PR because we were both modifying the same thing. Feel free to modify as needed!

psiemens replied:
Sweet! my bad -- didn't realize at first

psiemens commented on dispatch/static/manager/src/js/components/Header.js:
```diff
@@ -54,6 +54,9 @@ export default function Header(props) {
                 </li>
+                <li className='c-header__dropdown__link'>
+                  <Link to='/events/audit'><span className='pt-icon-standard pt-icon-endorsed' /> audit events</Link>
+                </li>
```
Instead of adding a separate link in the navigation, I think it would be better to link to this page from the main events listing. Something like this:

![screen shot 2017-06-13 at 9 16 06 am](https://user-images.githubusercontent.com/2547035/27084263-388cb77c-5019-11e7-90ec-4ef6c1d4f6a8.png)

Clicking on "5 pending events" would take the user to the approval page.

I used the Blueprint tag component with the following classNames:

```html
<div class="pt-tag pt-large pt-intent-warning">5 pending events</div>
```

psiemens commented on dispatch/static/manager/src/js/components/inputs/LinkButton.js:
```diff
@@ -5,7 +5,12 @@ import { Button } from '@blueprintjs/core'
+      to={props.to}
+      onClick={e => {
+        if (props.disabled) {
+          e.preventDefault()
```
Nice addition 👍

psiemens commented on dispatch/static/manager/src/js/index.js:
```diff
@@ -100,6 +100,7 @@ render((
           <Route path='events'>
             <IndexRoute component={Pages.Events.Index} />
             <Route path='new' component={Pages.Events.NewEvent} />
+            <Route path='audit' component={Pages.EventAudit} /> 
```
The audit page should be grouped with the other events pages and referenced like so:

`component={Pages.Events.Audit}`

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/EventAuditPagination.js:
```diff
@@ -0,0 +1,35 @@
+      <LinkButton
+        to={getPath(props.location, page - 1)}
+        disabled={page <= 1}>Prev</LinkButton>
+      {' '}
```
Use `margin` to add space between buttons rather than a hardcoded space

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/EventAuditPagination.js:
```diff
@@ -0,0 +1,35 @@
+    pathname: location.pathname,
+    query: query
+  }
+}
```
Is this done anywhere else in the application? Does it make sense to move to `util/helpers.js`?

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/EventCard.js:
```diff
@@ -0,0 +1,92 @@
+import { LinkButton } from '../../components/inputs'
+import { humanizeDatetime } from '../../util/helpers'
+
+function labelAndProp(label, prop) {
```
Convert this to a functional React component, then call like so below:

`<EventLabel label='Description' value={this.props.event.description} />`

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/EventCard.js:
```diff
@@ -0,0 +1,92 @@
+            transitionDuration={15}
+            content="Delete">
+            <AnchorButton
+              disabled={this.state.disableButtons}
```
Can you explain the need for `this.state.disableButtons`?

joncppl replied:
just to disable them during the animation.
Since I'll be removing that it won't be necessary anymore

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/index.js:
```diff
@@ -0,0 +1,181 @@
+    super(props)
+
+    this.state = {
+      count: 0
```
How hard would it be to control this in the store? I know it works this way, but it breaks out of the redux architecture and seems kinda hard to follow. I might be missing something though

Ideally the events list that we keep in the store would stay up to date, and any time an event goes from `is_submission=True` to `is_submission=False`, that change would be reflected. Then you can simply count up all of the events with `is_submission=True` and use that as your count

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/index.js:
```diff
@@ -0,0 +1,181 @@
+    const query = R.clone(PENDING_QUERY)
+
+    if (props.location.query.page) {
+      query.offset = (Number(props.location.query.page) - 1) * PER_PAGE
```
use `parseInt(props.location.query.page, 10)` instead of `Number`

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/index.js:
```diff
@@ -0,0 +1,181 @@
+    this.checkPage()
+    this.setState({ count: this.state.count - 1 })
+    this.reloadListFlag = false
+  }
```
I think it might be better to create a new action for this, and then just call `this.props.approveEvent(this.props.token, id)`

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/index.js:
```diff
@@ -0,0 +1,181 @@
+    this.checkPage()
+    this.setState({ count: this.state.count - 1 })
+    this.reloadListFlag = false
+  }
```
Same for disapproving events

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/index.js:
```diff
@@ -0,0 +1,181 @@
+
+  render() {
+    this.eventCount = 0
+    const events = this.props.events.ids.map((id, idx) => {
```
I prefer to chain this logic together like so:

```js
const events = this.props.events.ids
  .map(id => this.props.entities.events[id])
  .map((event, i) => {
    // ...
  })

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/index.js:
```diff
@@ -0,0 +1,181 @@
+
+      return (
+        <EventCard
+          key={idx}
```
Use `i` instead of `idx` (for consistency with rest of application)

psiemens commented on dispatch/static/manager/src/js/pages/EventAudit/index.js:
```diff
@@ -0,0 +1,181 @@
+                location={this.props.location} />: null}
+            {this.state.count} event{this.state.count == 1 ? '' : 's'} pending approval
+          </div>
+          <ReactCSSTransitionGroup
```
This transitions are nice, but I think it'd prefer to remove them for now since we don't use them anywhere else in the application

psiemens reviewed (changes requested):
This is really good overall. I think we might disagree on some of the code design choices, so feel free to argue your side.

I try to manage the applications state entirely from inside the store so that components deal with as little "business logic" as possible. There are times when that pattern should be broken, but I'm not sure this is one of them.

joncppl committed:
fixes

- move some code
- make event audit entirely redux/store driven
- remove animations

joncppl committed:
move event auditor out of header dropdown

instead add a pending events alert to the event manager

joncppl committed:
don't show alert when there are no pending events

joncppl committed:
remove custom approve action

instead use improve prop checking logic

joncppl committed:
change if to else if

psiemens committed:
Move pending button to left side of toolbar

psiemens committed:
Switch to generic pagination component

psiemens committed:
Update event auditor styles
